python/paddle/tensor/einsum.py

```diff
@@ -0,0 +1,909 @@
+
+
+# def bcastable_test(args, f=None):
+#     '''
+#     Tests if the two operands can perform a broadcast operation on the given ranges of dimensions. 
+#     We follow the Numpy broadcasting convention which states that, by lining up the shape arrays
+#     starting from the right most dimension, all the aligned dimensions either have equal sizes or
+#     one of them is sized one.
+#     Parameters
+#     ----------
+#     args:
+#         *args unpacks into operand one's axes range, shape, operand two's axes range, shape
+#     f: 
+#         if available, is used as a callback for postprocessing the aligned operand dimensions.
+#     '''
+#     xran, xshape, yran, yshape = args
+#
+#     xran_inv, yran_inv = xran[::-1], yran[::-1]
+#
+#     for xi, yi in zip(xran_inv, yran_inv):
+#         xs, ys = xshape[xi], yshape[yi]
+#         cond = xs == ys or xs == 1 or ys == 1
+#         if not cond:
+#             return False
+#
+#     if not f:
+#         return True
+#
+#     # Apply the callback to each aligned dimension pair
+#     for xi, yi in zip(xran_inv, yran_inv):
+#         f(xi, yi)
+
+
+def build_view(in_labels, out_labels):
+    '''
+    Build an inverse map of dimension indices. Three conditions must hold for 
+    the result to be meaningful. 
+    First, no duplicate letter labels in each label string.
+    Second, the number of dots in dimout_labels >= that in in_labels.
+    Third, dots are contiguous in each label string.
+
+    Parameters
+    ----------
+    in_labels:
+        The dimension labels to map to
+    out_labels:
+        The dimension labels to map from
+    
+    Returns
+    -------
+    The inverse map from out_labels to in_labels. The length of the inverse map equals that of
+    out_labels. -1 is filled if there's no matching intput dimension for a specific label.
+
+    Examples
+    --------
+    in_labels = 'ij..', out_labels = '..ji'
+    inv_map = [2, 3, 1, 0]
+    in_labels = 'ij..', out_labels = '..kji'
+    inv_map = [2, 3, -1, 1, 0]
+    '''
+    # print(f"in labels: '{in_labels}'     out labels: '{out_labels}'")
+
+    inv_map = [-1] * len(out_labels)
+
+    # First build the broadcast dimension mapping
+    # Find the broadcast index range in out_labels
+    r = re.search(r'\.+', out_labels)
+    if r:
+        start, end = r.start(), r.end()
+        s = re.search(r'\.+', in_labels)
+        # fill the broadcast dimension indices from right to left.
+        if s:
+            for ax, dim in zip(
+                    range(start, end)[::-1], range(s.start(), s.end())[::-1]):
+                inv_map[ax] = dim
+
+    # Now work on non-broadcast dimensions 
+    if r:
+        it = itertools.chain(range(start), range(end, len(out_labels)))
+    else:
+        it = iter(range(len(out_labels)))
+
+    for i in it:
+        inv_map[i] = in_labels.find(out_labels[i])
+
+    return inv_map
+
+
+# def part_labels(nop_labels, rhs, n_bcast_dims):
```

Comment:
Same as above. It can be removed for simplicity.

Reply:
OK